File: wordle/feedback.py
"""Feedback computation using two-pass algorithm with base-3 pattern encoding.

The pattern is encoded as a base-3 integer where each digit (MSB to LSB)
corresponds to a position: 0=GREY, 1=YELLOW, 2=GREEN.

For L=5, patterns range from 0 (all grey) to 242 (all green).

Two-pass algorithm:
  Pass 1: Mark greens (exact matches), decrement letter counts.
  Pass 2: Mark yellows left-to-right using remaining counts.
This correctly handles duplicate letters.
"""
from __future__ import annotations

_ORD_A = ord("a")


def compute_pattern(guess: bytes, secret: bytes, word_length: int = 5) -> int:
    """Compute the feedback pattern for guess against secret.

    Args:
        guess: The guessed word as lowercase ASCII bytes.
        secret: The secret word as lowercase ASCII bytes.
        word_length: Length of the words.

    Returns:
        Base-3 encoded pattern int in [0, 3^word_length).
    """
    counts = [0] * 26
    result = [0] * word_length

    # Count letters in secret
    for i in range(word_length):
        counts[secret[i] - _ORD_A] += 1

    # Pass 1: greens
    for i in range(word_length):
        if guess[i] == secret[i]:
            result[i] = 2
            counts[guess[i] - _ORD_A] -= 1

    # Pass 2: yellows
    for i in range(word_length):
        if result[i] == 0:
            idx = guess[i] - _ORD_A
            if counts[idx] > 0:
                result[i] = 1
                counts[idx] -= 1

    # Encode as base-3 integer (MSB = position 0)
    pattern = 0
    for v in result:
        pattern = pattern * 3 + v
    return pattern


def compute_pattern_batch(
    guess: bytes,
    candidates: list[bytes],
    word_length: int = 5,
) -> list[int]:
    """Compute patterns for one guess against many candidates."""
    a = _ORD_A
    patterns = []
    for secret in candidates:
        counts = [0] * 26
        result = [0] * word_length
        for i in range(word_length):
            counts[secret[i] - a] += 1
        for i in range(word_length):
            if guess[i] == secret[i]:
                result[i] = 2
                counts[guess[i] - a] -= 1
        for i in range(word_length):
            if result[i] == 0:
                idx = guess[i] - a
                if counts[idx] > 0:
                    result[i] = 1
                    counts[idx] -= 1
        p = 0
        for v in result:
            p = p * 3 + v
        patterns.append(p)
    return patterns


def bucket_sizes(
    guess: bytes,
    candidates: list[bytes],
    word_length: int = 5,
) -> dict[int, int]:
    """Partition candidates by pattern and return counts only.

    More memory-efficient than bucket_by_pattern for scoring.
    """
    a = _ORD_A
    buckets: dict[int, int] = {}
    for secret in candidates:
        counts = [0] * 26
        result = [0] * word_length
        for i in range(word_length):
            counts[secret[i] - a] += 1
        for i in range(word_length):
            if guess[i] == secret[i]:
                result[i] = 2
                counts[guess[i] - a] -= 1
        for i in range(word_length):
            if result[i] == 0:
                idx = guess[i] - a
                if counts[idx] > 0:
                    result[i] = 1
                    counts[idx] -= 1
        p = 0
        for v in result:
            p = p * 3 + v
        buckets[p] = buckets.get(p, 0) + 1
    return buckets


def bucket_by_pattern(
    guess: bytes,
    candidates: list[bytes],
    word_length: int = 5,
) -> dict[int, list[bytes]]:
    """Partition candidates into buckets keyed by feedback pattern."""
    a = _ORD_A
    buckets: dict[int, list[bytes]] = {}
    for secret in candidates:
        counts = [0] * 26
        result = [0] * word_length
        for i in range(word_length):
            counts[secret[i] - a] += 1
        for i in range(word_length):
            if guess[i] == secret[i]:
                result[i] = 2
                counts[guess[i] - a] -= 1
        for i in range(word_length):
            if result[i] == 0:
                idx = guess[i] - a
                if counts[idx] > 0:
                    result[i] = 1
                    counts[idx] -= 1
        p = 0
        for v in result:
            p = p * 3 + v
        if p in buckets:
            buckets[p].append(secret)
        else:
            buckets[p] = [secret]
    return buckets


def filter_candidates(
    candidates: list[bytes],
    guess: bytes,
    pattern: int,
    word_length: int = 5,
) -> list[bytes]:
    """Return candidates consistent with the given guess/pattern feedback."""
    a = _ORD_A
    result_list = []
    for secret in candidates:
        counts = [0] * 26
        result = [0] * word_length
        for i in range(word_length):
            counts[secret[i] - a] += 1
        for i in range(word_length):
            if guess[i] == secret[i]:
                result[i] = 2
                counts[guess[i] - a] -= 1
        for i in range(word_length):
            if result[i] == 0:
                idx = guess[i] - a
                if counts[idx] > 0:
                    result[i] = 1
                    counts[idx] -= 1
        p = 0
        for v in result:
            p = p * 3 + v
        if p == pattern:
            result_list.append(secret)
    return result_list

File: src/solver/feedback.js
/**
 * Wordle feedback computation using two-pass algorithm.
 *
 * Pattern is a base-3 integer (0-242) where each digit is:
 *   0 = grey, 1 = yellow, 2 = green
 *
 * Two-pass algorithm:
 *   Pass 1: Mark greens (exact matches), decrement letter counts.
 *   Pass 2: Mark yellows left-to-right using remaining counts.
 */

const A = 97; // 'a'.charCodeAt(0)

/** All-green pattern = 3^5 - 1 = 242 */
export const ALL_GREEN = 242;

/**
 * Compute feedback pattern for a guess against a secret.
 * @param {string} guess - 5-letter guess
 * @param {string} secret - 5-letter secret
 * @returns {number} Base-3 pattern (0-242)
 */
export function computePattern(guess, secret) {
  const counts = new Int8Array(26);
  const result = new Int8Array(5);

  for (let i = 0; i < 5; i++) {
    counts[secret.charCodeAt(i) - A]++;
  }

  // Pass 1: greens
  for (let i = 0; i < 5; i++) {
    if (guess.charCodeAt(i) === secret.charCodeAt(i)) {
      result[i] = 2;
      counts[guess.charCodeAt(i) - A]--;
    }
  }

  // Pass 2: yellows
  for (let i = 0; i < 5; i++) {
    if (result[i] === 0) {
      const idx = guess.charCodeAt(i) - A;
      if (counts[idx] > 0) {
        result[i] = 1;
        counts[idx]--;
      }
    }
  }

  return result[0] * 81 + result[1] * 27 + result[2] * 9 + result[3] * 3 + result[4];
}

/**
 * Get bucket sizes (pattern -> count) for a guess against candidates.
 * @param {string} guess
 * @param {string[]} candidates
 * @returns {Map<number, number>}
 */
export function bucketSizes(guess, candidates) {
  const buckets = new Map();
  const counts = new Int8Array(26);
  const result = new Int8Array(5);
  const gc = [
    guess.charCodeAt(0) - A,
    guess.charCodeAt(1) - A,
    guess.charCodeAt(2) - A,
    guess.charCodeAt(3) - A,
    guess.charCodeAt(4) - A,
  ];

  for (let c = 0; c < candidates.length; c++) {
    const secret = candidates[c];
    counts.fill(0);
    result.fill(0);

    for (let i = 0; i < 5; i++) {
      counts[secret.charCodeAt(i) - A]++;
    }
    for (let i = 0; i < 5; i++) {
      if (gc[i] === secret.charCodeAt(i) - A) {
        result[i] = 2;
        counts[gc[i]]--;
      }
    }
    for (let i = 0; i < 5; i++) {
      if (result[i] === 0 && counts[gc[i]] > 0) {
        result[i] = 1;
        counts[gc[i]]--;
      }
    }

    const p = result[0] * 81 + result[1] * 27 + result[2] * 9 + result[3] * 3 + result[4];
    buckets.set(p, (buckets.get(p) || 0) + 1);
  }

  return buckets;
}

/**
 * Filter candidates to those consistent with a guess/pattern.
 * @param {string[]} candidates
 * @param {string} guess
 * @param {number} pattern
 * @returns {string[]}
 */
export function filterCandidates(candidates, guess, pattern) {
  const result = [];
  for (let i = 0; i < candidates.length; i++) {
    if (computePattern(guess, candidates[i]) === pattern) {
      result.push(candidates[i]);
    }
  }
  return result;
}

/**
 * Convert pattern to emoji string.
 * @param {number} pattern
 * @returns {string}
 */
export function patternToEmoji(pattern) {
  const tiles = ['', '', '', '', ''];
  const map = ['\u2b1b', '\ud83d\udfe8', '\ud83d\udfe9'];
  for (let i = 4; i >= 0; i--) {
    tiles[i] = map[pattern % 3];
    pattern = (pattern / 3) | 0;
  }
  return tiles.join('');
}

/**
 * Convert pattern to compact string like "00120".
 * @param {number} pattern
 * @returns {string}
 */
export function patternToCompact(pattern) {
  const digits = ['', '', '', '', ''];
  for (let i = 4; i >= 0; i--) {
    digits[i] = String(pattern % 3);
    pattern = (pattern / 3) | 0;
  }
  return digits.join('');
}
